### PR TITLE
feat: add Gemini CLI usage tracking via gemistat

### PR DESCRIPTION
**Pull request by nihalnihalani**
## Summary

- **Adds Gemini CLI as the third usage source** alongside Claude Code (ccusage) and Codex (@ccusage/codex), using [gemistat](https://github.com/ryoppippi/gemistat) by ryoppippi (same author as ccusage)
- **Mirrors the existing Codex integration pattern exactly** — subprocess call, JSON parsing, silent failure, 3-way date merge
- **Gemini models display with Google brand colors** across CLI scorecard and web feed, with proper prettification of all model IDs

### CLI Core (`packages/cli/`)

| File | Change |
|------|--------|
| `src/lib/gemini.ts` | **New** — gemistat subprocess runner (`bunx gemistat@1 daily --json`), JSON parser, `toIsoDate()` date converter (gemistat expects `YYYY-MM-DD`, not `YYYYMMDD`) |
| `src/commands/push.ts` | 3-way `mergeEntries(claude, codex, gemini)`, gemini in `Promise.all`, anomaly/blocked-date filtering, hash includes gemini raw |
| `src/lib/token-normalization.ts` | Added `"gemini"` to `TokenSourceHints.source` union |
| `src/lib/ccusage.ts` | Added `"gemini"` to `NormalizationAnomaly.source` union |
| `src/components/theme.ts` | 8 Gemini model colors (Google Blue → Amber gradient by tier) |
| `src/components/ModelPalette.tsx` | Gemini `prettifyModel` + `getModelColor` fallback |

### Web (`apps/web/`)

| File | Change |
|------|--------|
| `components/app/feed/ActivityCard.tsx` | Gemini `prettifyModel`, suffix stripping, 8 Gemini entries + fallback in `modelColor()` |
| `lib/utils/post-share.ts` | Gemini `prettifyModel` + suffix stripping |
| `lib/share-assets/post-card-image.tsx` | Gemini `prettifyModel` + suffix stripping |
| `lib/open-stats.ts` | Gemini `prettifyModel` + suffix stripping |

### Tests (20 new tests)

| File | Tests |
|------|-------|
| `__tests__/gemini.test.ts` | **New** — 14 tests: parsing, date conversion, multi-model, silent failure, edge cases |
| `__tests__/merge.test.ts` | **New** — 3 tests: 3-way merge, gemini-only, backward compat |
| `__tests__/commands/push.test.ts` | 3 new: timeout forwarding, silent failure, gemini-only submission |
| `apps/web/__tests__/unit/prettify-model.test.ts` | 8 new Gemini model prettification cases |
| `__tests__/flows/cli-sync-flow.test.ts` | Updated for 3rd subprocess mock |

### Docs

- `CHANGELOG.md` — Gemini integration entry under Unreleased
- `DECISIONS.md` — Architecture decision: gemistat subprocess vs direct telemetry parsing
- `docs/plans/2026-04-04-gemini-cli-integration.md` — Full implementation plan

## Design decisions

**Why gemistat (not direct telemetry parsing)?**
- Same author as ccusage, same output format (`{ daily: [...], totals }`)
- Handles OpenTelemetry parsing, LiteLLM pricing lookup, and daily aggregation
- Keeps the CLI thin — no Gemini pricing table to maintain
- Silent failure like Codex — existing users unaffected

**Key details:**
- Pinned to `gemistat@1` (major version, avoids registry roundtrip)
- Dates converted via `toIsoDate()` — gemistat expects `YYYY-MM-DD`, ccusage expects `YYYYMMDD`
- `totalTokens` computed client-side (gemistat doesn't provide it)
- Cache semantics: `"separate"` (same as ccusage)
- Preview/experimental suffixes stripped: `gemini-3.1-pro-preview-05-06` → `Gemini 3.1 Pro`

## Data flow

```
~/.gemini/telemetry.log
    ↓
gemistat (bunx gemistat@1 daily --json)
    ↓
gemini.ts: parseGeminiOutput() → CcusageDailyEntry[]
    ↓
push.ts: mergeEntries(claude, codex, gemini)
    ↓
POST /api/usage/submit (no server changes needed)
```

## Test plan

- [x] CLI: 225 tests pass (16 files)
- [x] Web: 20 prettify-model tests pass
- [x] TypeScript: compiles cleanly (`tsc` in CLI package)
- [ ] Manual: install gemistat, enable Gemini CLI telemetry, run `straude push`, verify Gemini models appear in feed
- [ ] Manual: run `straude push` without gemistat installed — verify silent failure, Claude/Codex data still submits
- [ ] Manual: verify Gemini model colors render correctly in CLI scorecard and web feed

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added Gemini usage tracking to the CLI and unified Gemini model display with human-readable names and dedicated palette across CLI and web.

* **Bug Fixes**
  * Fixed inconsistent Gemini model name formatting and adjusted fallback messaging to reference “other sources” when local data is missing.

* **Tests**
  * Added comprehensive tests for Gemini parsing, merging, subprocess behavior, and push workflows.

* **Documentation**
  * Updated changelog, decisions, and integration plan for Gemini support.

### CODE REVIEW

File: apps/web/components/app/feed/ActivityCard.tsx [hunk 1]
```diff
@@ -171,6 +183,15 @@ function modelColor(name: string): string {
...
   if (/^o4/i.test(name)) return "#6366F1";
+  if (/Gemini 3\.1 Pro/.test(name)) return "#4285F4";
+  if (/Gemini 3\.1 Flash Lite/.test(name)) return "#00ACC1";
+  if (/Gemini 3 Flash/.test(name)) return "#009688";
+  if (/Gemini 2\.5 Pro/.test(name)) return "#3F51B5";
+  if (/Gemini 2\.5 Flash Lite/.test(name)) return "#8BC34A";
+  if (/Gemini 2\.5 Flash/.test(name)) return "#34A853";
+  if (/Gemini 2\.0 Flash Lite/.test(name)) return "#FF8F00";
+  if (/Gemini 2\.0 Flash/.test(name)) return "#FBBC05";
+  if (/Gemini/i.test(name)) return "#3F51B5"; // Google indigo fallback
```
[the remark this answers is not in the document]
Comment (nihalnihalani) [reply apps/web/components/app/feed/ActivityCard.tsx:194]: This is intentional and consistent with the existing pattern. The `modelColor()` function is a data-visualization palette that maps brand-representative hex colors to each AI model for the usage breakdown bar. **All** models in this function use hardcoded hex values — Claude (`#DF561F`, `#F08A5D`, `#F7B267`), GPT (`#2A9D8F`, `#4C78A8`), o3/o4 (`#3B82F6`, `#6366F1`), and the 6-color fallback palette on line 196. The Gemini entries follow the exact same established pattern.

The theme tokens in `globals.css` `@theme inline` define UI chrome colors (background, foreground, accent, border, muted, subtle, etc.) — there are no model-specific color tokens. Moving only the Gemini colors into the theme while leaving Claude/GPT/o3/o4 hardcoded would be inconsistent. Additionally, the function uses regex-based matching (not exact string equality), so a simple constant map wouldn't be a drop-in replacement without refactoring the matching logic.

No change needed here.

File: packages/cli/src/lib/gemini.ts [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+      encoding: "utf-8",
+      timeout: timeoutMs ?? DEFAULT_SUBPROCESS_TIMEOUT_MS,
+      maxBuffer: 10 * 1024 * 1024,
+      shell: process.platform === "win32",
+    }, (err, stdout) => {
+      if (err) reject(err);
+      else resolve(stdout);
+    });
+  });
+}
```
[the remark this answers is not in the document]
Comment (nihalnihalani) [reply packages/cli/src/lib/gemini.ts:77]: Good catch — fixed in 925d96f. Changed `GEMISTAT_PKG` from `"gemistat@1"` to `"gemistat@0"`. The npm registry only has 0.x versions (0.0.0, 0.1.0, 0.1.1), so the `@1` range could never resolve.

**Review by ohong — COMMENTED** [review]
Nightly review (automated). Flagging up front: **no GitHub Actions CI run exists for this PR at all** — it looks blocked pending first-time-contributor workflow approval. The "Vercel — failure" status is just the external-fork deploy-authorization gate, not a real build failure. This PR cannot be called ready to merge regardless of code quality until CI actually runs.

**Blocking structural issue**: this PR's base (`d435ab1`) is ~175 commits behind current `main`, and `main`'s usage-collection architecture has since changed substantially (unified `ccusage v20` collector replacing the standalone `codex.ts`/token-normalizer this PR's `gemini.ts` mirrors — see the "Unified usage collection on ccusage v20" CHANGELOG entry). This isn't a simple rebase conflict; the Gemini integration will need to be re-verified against the new collector pattern after rebasing onto current `main`.

**Correctness**: date handling (`gemistat`'s `YYYY-MM-DD` vs `ccusage`'s `YYYYMMDD`) looks correctly converted and the 3-way merge keys align. Package pin was fixed from the non-existent `gemistat@1` to the real `gemistat@0` in a follow-up commit — worth double-checking that resolves cleanly post-rebase.

**Security**: subprocess invocation mirrors the existing Codex pattern (array args, no shell interpolation) — no new injection surface introduced.

Also worth Oscar's attention: this overlaps with #22, which adds Gemini tracking via a different mechanism (`gcusage` + session-file fallback vs. this PR's `gemistat`). Only one Gemini integration approach should land.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by ohong** [comment]
Thanks for this, and sorry it sat so long. The feature you're asking for is one we want — but the CLI was rewritten underneath this branch, and the rewrite makes most of the diff unnecessary. Wanted to explain rather than just close it.

**Gemini usage already flows end to end on `main` today.** The CLI now shells out to `ccusage` as a single unified collector rather than parsing agent logs itself. `packages/cli/src/lib/ccusage.ts:454` builds `["daily", "--json", "--since", ..., "--until", ...]` with no source scoping, and `ccusage daily` includes *every* source it detects. The pinned version (`ccusage@^20.0.16`) natively supports Gemini CLI, Qwen, Kimi, Copilot, Amp, Droid, Goose, OpenCode and others.

Nothing downstream is hardcoded to Claude and Codex either:

- `CcusageAgent` is `string`, and agent names are read straight from `metadata.agents` per row (`ccusage.ts:297`).
- The submit endpoint validates that `ccusage_agents` entries are non-empty strings and nothing more — no allowlist (`apps/web/app/api/usage/submit/route.ts:74-81`).
- Unknown model names now get a stable colour from a hash-derived palette instead of falling through, so Gemini models render correctly on activity cards (`apps/web/lib/constants/model-colors.ts`).

The only Claude/Codex-specific code left is two cosmetic collector labels in `collectorForAgents` (`ccusage.ts:348-349`).

So the `gemistat` dependency, the custom parser in `packages/cli/src/lib/gemini.ts`, and the token-normalization layer are all solving a problem the ccusage collector already solves — and solving it a second way would mean two pricing paths and two dedup paths on a financial code path, which is where we least want divergence.

**What's actually still open**, and worth a much smaller PR if you're up for it:

1. `prettifyModel` handling for Gemini model IDs, so they read as "Gemini 2.5 Pro" rather than the raw string. Your `apps/web/__tests__/unit/prettify-model.test.ts` is directly reusable here.
2. Explicit Gemini entries in `MODEL_COLOR_PATTERNS` so the family gets a deliberate colour instead of a hashed one.
3. Docs saying which agents are picked up, since "install the Gemini CLI and it just works" isn't discoverable anywhere right now.

If you'd rather not rework it, no problem at all — say the word and I'll close this and open a scoped issue crediting you for the report. There's also #22 covering overlapping ground, so whatever lands should cover both.

**Comment by ohong** [comment]
Thanks @nihalnihalani for suggesting Gemini CLI support and sharing this implementation. We completed the remaining work in #156 using Straude's unified ccusage integration.

Gemini already flows through that collector. We raised the dependency floor to ccusage 20.0.20, which covers all 16 agents in the current published release, including Gemini CLI, Qwen, and Grok Build CLI. Real-binary tests now cover Gemini-only, Qwen-only, Grok-only, and mixed usage. Push/API tests preserve source metadata, and the public docs describe the broader support. Join/share pages also stop incorrectly labeling other agents as Claude Code or Codex.

Keeping this in ccusage gives us one parsing, pricing, and deduplication path, so the separate gemistat adapter is no longer needed. The change passed 217 CLI tests, 747 web tests, typechecking, the production build, a clean packed-install sync, and independent code review.

The code is merged in #156. Full [CI](https://github.com/ohong/straude/actions/runs/33932389627), including real database integration and browser tests, passed. Straude 0.1.31 is prepared; npm publication is still pending renewed authentication.

Closing this as superseded by #156. Thank you again for the suggestion and the work you put into it!
